ihmc-java-toolkit/src/main/java/us/ihmc/tools/thread/LoopingThread.java

```diff
@@ -0,0 +1,193 @@
+   }
+
+   /**
+    * Starts executing the {@link #runInLoop()} method in a loop.
+    */
+   @Override
+   public void start()
+   {
+      synchronized (runCommand)
+      {
+         initializeIfNeeded();
+
+         runCommand.set(RunCommand.LOOP);
+         runCommand.notify();
+      }
+   }
+
+   /**
+    * Signal the thread to execute the {@link #runInLoop()} method once, then pause.
+    */
+   public void runOnce()
+   {
+      synchronized (runCommand)
+      {
+         initializeIfNeeded();
+
+         runCommand.set(RunCommand.RUN_ONCE);
+         runCommand.notify();
+      }
+   }
+
+   private void initializeIfNeeded()
+   {
+      if (runCommand.compareAndSet(RunCommand.INITIALIZE, RunCommand.PAUSE))
+         super.start();
+   }
+
+   /**
+    * Pauses re-executing the {@link #runInLoop()} method upon its completion.
+    */
+   public void pause()
+   {
+      runCommand.set(RunCommand.PAUSE);
+   }
+
+   /**
+    * Finalizes the execution of {@link #runInLoop()} and stops the thread.
+    * The thread cannot be re-started after calling this method.
+    */
+   public void close()
+   {
+      synchronized (runCommand)
+      {
+         runCommand.set(RunCommand.FINALIZE);
+         runCommand.notify();
+      }
+   }
+
+   /**
+    * Signals the thread to finalize execution of {@link #runInLoop()}
+    * and waits until the thread exits.
+    * Same as calling {@link #close()} then {@link #join()}.
+    */
+   public void blockingClose()
+   {
+      close();
+      try
+      {
+         join();
+      }
+      catch (InterruptedException ignored) {}
+   }
+
+   /**
+    * The method that is executed in a loop.
+    * You may either provide this class a
+    * {@link Runnable}/{@link RunnableThatThrows} to run,
+    * or override this method.
+    */
+   public void runInLoop()
+   {
+      if (target != null)
+         target.run();
+   }
+
+   @Override
+   public final void run()
+   {  // Run while the thread is not finalizing
+      while (runCommand.get() != RunCommand.FINALIZE)
+      {
+         synchronized (runCommand)
+         {
+            RunCommand currentCommand = runCommand.get();
+            if (currentCommand == RunCommand.PAUSE)
+            {  // Currently paused
+               try
+               {  // Wait until we're notified to run
+                  runCommand.wait();
+               }
+               catch (InterruptedException ignored) {}
+               continue;
+            }
+            else if (currentCommand == RunCommand.RUN_ONCE)
+               pause(); // Set the run command to pause after this loop
+         }
+
+         if (loopPeriodLimit > 0.0)
+            throttler.waitAndRun(loopPeriodLimit);
+
+         runInLoop();
+      }
+   }
+
+   public boolean isLooping()
+   {
+      return runCommand.get() == RunCommand.LOOP;
+   }
+
+   private enum RunCommand
+   {
+      INITIALIZE, LOOP, PAUSE, RUN_ONCE, FINALIZE
```

Comment:
Can we remove INITIALIZE and use PAUSE at the beginning?

Reply:
Why do you need to delay the creation of the thread? Can't you do it on construction?

Reply:
You're right, I could. I was trying to keep the semantics of `start()` similar to `Thread`, where it would create the thread when `start()` is called. But I suppose there's no harm in creating the thread upon construction.